### PR TITLE
refactor(color): classes used to manage models and labels, improve duplication UX

### PR DESCRIPTION
**Pull request by philmoz**
Refactor:
- Rework the classes used to manage models and labels to simplify the interface. The interface is reduced from 3 classes (ModelCell, ModelMap and ModelsList) to 2 by removing the ModelMap class and moving the functionality to the other two. The primary interface object has been renamed modelCellManager to make to code more readable.
- Use storageDirty to mark the labels file for update instead of saving immediately.
- Remove the delete / rebuild of the model management class objects when labels are renamed or deleted. Keeps the ModelCell pointers valid so the manage models page does not need to be rebuilt.

Enhancements:
- prevent invalid characters from being entered with the keyboard when creating labels (instead of deleting after entering the label).
- when a model is duplicated a unique name is set on the new model based on the current name and focus is set to the new model button in the manage models UI.

### CODE REVIEW

**Comment by pfeerick** [comment]
> * when a model is duplicated revert the name of the new model to the default name (MODELxx) and set focus to the new model button in the manage models UI.

This is because it would be harder to figure out what numerical number to add to the end to differentiate? As I don't see having to re-enter "Skyhunter" a second time if I duplicate the model as an enhancement (let alone anything more complex naming). If so, I get it, but it certainly isn't an _enhancement_.

**Comment by philmoz** [comment]
> > * when a model is duplicated revert the name of the new model to the default name (MODELxx) and set focus to the new model button in the manage models UI.
> 
> This is because it would be harder to figure out what numerical number to add to the end to differentiate? As I don't see having to re-enter "Skyhunter" a second time if I duplicate the model as an enhancement (let alone anything more complex naming). If so, I get it, but it certainly isn't an _enhancement_.

Well I thought it would be better than having multiple models all with the same name; but yes it would be better to try and append a unique number to the current name.

If there is no space for a unique number would it be better to fall back to the default name or truncate the current name to make room?

**Comment by pfeerick** [comment]
> Well I thought it would be better than having multiple models all with the same name; 

I am in full agreement on that!

but yes it would be better to try and append a unique number to the current name.
> 
> If there is no space for a unique number would it be better to fall back to the default name or truncate the current name to make room?

The easy option is to fall back, but truncation preserves as much of the original name as possible. It may even be worth considering splitting that behaviour out of this PR, as ideally you want bw and colorlcd to behave the same way, so may require changes there to keep from introducing diverging behaviours (in a PR ostensibly focused on colorlcd refactor)?

**Comment by philmoz** [comment]
> The easy option is to fall back, but truncation preserves as much of the original name as possible. It may even be worth considering splitting that behaviour out of this PR, as ideally you want bw and colorlcd to behave the same way, so may require changes there to keep from introducing diverging behaviours (in a PR ostensibly focused on colorlcd refactor)?

The issue with B&W is the model duplication is done by just physically copying the model yaml file.
The rename is possible now with color radios because of the ability to rewrite the model file without mangling the active model data. This isn't an option for B&W at the moment.

**Comment by philmoz** [comment]
Updated to try and generate a unique name from the current name by appending a number as "(x)" to the current name.
If the current name is too long it is truncated and if a unique name can't be created it falls back to the default.

**Comment by pfeerick** [comment]
Couple of things:
- `renameLabel()` doesn't have any input validation any more... so among other things you can rename a label using prohibited characters, such as `"`... might want to review what other checksfrom ModelsList::renameLabel that are still relevant here
- shouldn't `textedit.cpp:79` be using `<=` ?? - it is dropping `~` even though though it it's not in the excluded character list per `labelExcludedChars`
- modelslist.cpp#L1213 doesn't seem to be used for anything (also on next line)?
- modelslist.cpp#L333 return value from is being swallowed - don't you want to show an error or something if that fails?

Claude is moaning about a off by one error, which will occur if someone is daft enough to duplicate a model with 12 character name more than 9 times... but it is still a off by one error regardless :shrug:

> modelslist.cpp:273.
> LEN_MODEL_NAME is 15 on colorlcd, and the buffer s is 16 bytes. For a name of exactly 12 characters, truncate is false. From the 10th duplicate on (i >= 10), "(10)" is written at s+12, which puts its NUL at index 16 (one past the buffer). The following strAppend(modelName, s) has no length bound, so it overflows modelName[16] as well.
> 
> Fix: decide truncation per iteration, using LEN_MODEL_NAME - (i >= 10 ? 4 : 3). Also pass LEN_MODEL_NAME to strAppend.
> Names of 13–15 characters lose 3 characters even when 1 would do. That is cosmetic.
> The setDefaultName() fallback isn't checked for uniqueness. The name check is also case-sensitive, while sorting is case-insensitive.

Added some more unit tests... they should fail until some of those regressions / edge cases are fixed. Let me know if any were over-eager or wrong. Only two are testing against new behaviours - four of the failures would have passed on the prior API - RejectedLabelDoesNotMakeModelLabelled and UniqueNameTwelveCharNameTenthDuplicate are the new edge case ones.

**Comment by philmoz** [comment]
@pfeerick Are the 'RenameLabelRejectsYamlUnsafeCharacters' and 'RenameLabelTruncatesToLabelLength' tests really needed?
Both of these cases are handled in the UI by the keyboard entry for the label name (length limit and invalid character rejection).

**Comment by philmoz** [comment]
@pfeerick The issues should be fixed. I've also fixed a number of issues with the maximum length of labels for a model in both companion and firmware.

File: radio/src/tests/modelslist_manager.cpp [hunk 1]
```diff
@@ -413,7 +413,7 @@ TEST_F(ModelCellManagerFsTest, LabelsMapMoveUpSwapsMembership)
 
 TEST_F(ModelCellManagerFsTest, RejectedLabelDoesNotMakeModelLabelled)
 {
-  ModelCell* cell = addCell("model0001.yml", "One");
```
Comment (pfeerick) [inline radio/src/tests/modelslist_manager.cpp]: Why this change? Testing against the default name (name, not label) is not relevant here.

File: radio/src/storage/modelslist.cpp [hunk 2]
```diff
@@ -52,674 +37,394 @@ using std::list;
...
+
+  _isDirty = false;
 }
 
-/**
- * @brief Adds a label to the filter, used in yaml_labelslist on load
- *
- * @param label Label to be added
- */
+void ModelCell::updateFinfoHash()
```
Comment (pfeerick) [inline radio/src/storage/modelslist.cpp:369]: This should either be marked as a TODO or removed, not just left disabled.

File: radio/src/storage/modelslist.cpp [hunk 3]
```diff
@@ -1450,3 +1022,418 @@ uint8_t ModelsList::findNextUnusedModelId(uint8_t moduleIdx)
...
+
+/**
+ * @brief Update a label string
+ *
+ * @param from label to rename
+ * @param to new label string
+ * @param progress function to update progress bar (if needed)
+ */
+
+void ModelsList::renameLabel(const std::string &from, const std::string& _to,
```
Comment (pfeerick) [inline radio/src/storage/modelslist.cpp:1237]: Why not check for fit with a local copy _before_ applying? Then wouldn't need the labels[idx] = to, the restore, or the second progress pass, or the deduplication step needed in getLabels because of this (it could stay, but is really a "shouldn't happen" rather than a "ignore duplicates" protection then. Plus would only run on models that have a label, instead of all of them.

**Review by pfeerick — COMMENTED** [review]
>Are the 'RenameLabelRejectsYamlUnsafeCharacters' and 'RenameLabelTruncatesToLabelLength' tests really needed?
Both of these cases are handled in the UI by the keyboard entry for the label name (length limit and invalid character rejection).

Yes, because the storage API should not blindly depend on the UI to sanitise it's input. Hence why these wouldn't have failed before if they had existed with the prior API, and failed now since the UI did *not* do that. If Lua API is extended to allow for adding / renaming labels, or other changes are made to the UI, this will make it so this can't happen again. 

Regarding e04c94a:

The old LABELS_LENGTH - 1 limit reserved the NUL (written by strAppend, etc).  The yaml copy_string() doesn't NUL terminate, so fromCSV() reads on into moduleData. With an internal module configured, the model's last label is dropped. The fact the radios writes it back itself from memory is just dumb luck. For Companion, BoundedString<N> excludes the NUL, so MODEL_LABELS_LEN 99 is probably correct?

File: radio/src/tests/modelslist_manager.cpp (hunk 1 above)
Comment (philmoz) [reply radio/src/tests/modelslist_manager.cpp]: It was failing the test; but that was something else. Will fix.

File: radio/src/storage/modelslist.cpp (hunk 2 above)
Comment (philmoz) [reply radio/src/storage/modelslist.cpp:369]: The comment in the function describes why.

File: radio/src/storage/modelslist.cpp (hunk 3 above)
Comment (philmoz) [reply radio/src/storage/modelslist.cpp:1237]: Don't follow this one. labels[idx] has.to be updated in order for the correct label string to be saved to current model or written back to other models and the labels.yml file. Doing it first means less messing around when testing for overflow (always using correct labels instead of applying an offset). The two restore cases are the exceptions so don't normally happen. Checking for duplicates in getLabels() is a safety net.

**Comment by philmoz** [comment]
LABELS_LENGTH set to 99 to match actual max length and space added to g_model buffer for nul terminator.
Fixed the 'RejectedLabelDoesNotMakeModelLabelled' test.
Replaced an unsafe call to strAppend in writeModelLabels.

File: radio/src/storage/modelslist.cpp (hunk 2 above)
Comment (pfeerick) [reply radio/src/storage/modelslist.cpp:369]: What is the basis of the compatibility issue? Companion does not care about it. The labels.yml format doesn't appear to be changed. This only seems to alter the `yaml_labelslist` behaviour... which could be a problem if an older version of the firmware was to read a newer file (although, that is just one of many). 

Perhaps the TODO is that a semver or cache version check is needed before this is enabled? i.e. 

```suggestion
// TODO: Enable once labels.yml records the version that wrote it and
// cached model data is discarded on a version mismatch.
void ModelCell::updateFinfoHash()
```

Meaning making it a 3.0 or other major version TODO, since prior versions would still be unprotected from changes to the cache structure, which is also not a new risk.

File: radio/src/storage/modelslist.cpp (hunk 3 above)
Comment (pfeerick) [reply radio/src/storage/modelslist.cpp:1237]: Something like 1a53310 . Progress use-after-free regression that LVGL memory allocator was masking was cleaned up in the commit before.

File: radio/src/storage/modelslist.cpp (hunk 2 above)
Comment (philmoz) [reply radio/src/storage/modelslist.cpp:369]: I chose to keep the existing functionality as recalculating the hash requires rewriting the labels.yml file as well.
I did not want to add this extra overhead without more thorough testing and analysis; but chose to leave the code in case it does get revisited in future.

File: radio/src/storage/modelslist.cpp [hunk 4]
```diff
@@ -1246,42 +1246,39 @@ void ModelsList::renameLabel(const std::string &from, const std::string& _to,
...
-
-      // Check that new label fits in all models
+      // Check that new label fits in all models using it
       for (auto it = begin(); it != end(); ++it) {
-        int newLen = toCSV((*it)->getLabels()).size();
-        if (newLen > LABELS_LENGTH) {
+        if (!(*it)->hasLabel(idx)) continue;
+        LabelsVector lbls = (*it)->getLabels();
+        lbls.erase(std::remove(lbls.begin(), lbls.end(), from), lbls.end());
+        if (toIdx < 0 || !(*it)->hasLabel(toIdx)) lbls.push_back(to);
```
Comment (philmoz) [inline radio/src/storage/modelslist.cpp]: For every affected model you are now copying the entire label vector and modifying it. This is much less efficient than restoring the label for the two (far less likely) cases of exceeding the limit or duplicating a label.

File: radio/src/storage/modelslist.cpp (hunk 3 above)
Comment (pfeerick) [reply radio/src/storage/modelslist.cpp:1237]: Better with f78ab2d ? This should still be reducing CPU cycles over the original form since it is skipping progress updates and models without the label in question.
